x/protorev/keeper/posthandler.go

```diff
@@ -0,0 +1,121 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	gammtypes "github.com/osmosis-labs/osmosis/v13/x/gamm/types"
+)
+
+type SwapToBackrun struct {
+	PoolId        uint64
+	TokenOutDenom string
+	TokenInDenom  string
+}
+
+type ProtoRevDecorator struct {
+	ProtoRevKeeper Keeper
+}
+
+func NewProtoRevDecorator(protoRevDecorator Keeper) ProtoRevDecorator {
+	return ProtoRevDecorator{
+		ProtoRevKeeper: protoRevDecorator,
+	}
+}
+
+// This posthandler will first check if there were any swaps in the tx. If so, collect all of the pools, build three
+// pool routes for cyclic arbitrage, and then execute the optimal route if it exists.
+func (protoRevDec ProtoRevDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// Create a cache context to execute the posthandler such that
+	// 1. If there is an error, then the cache context is discarded
+	// 2. If there is no error, then the cache context is written to the main context
+	cacheCtx, write := ctx.CacheContext()
+
+	txGasWanted := cacheCtx.GasMeter().Limit()
+	// Ignore cases where limit is 0 (edge case for genUtil init)
+	if txGasWanted == 0 {
+		return next(ctx, tx, simulate)
+	}
+
+	// Change the gas meter to be an infinite gas meter which allows the entire posthandler to work without consuming any gas
+	cacheCtx = cacheCtx.WithGasMeter(sdk.NewInfiniteGasMeter())
+
+	// Only execute the protoRev module if it is enabled
+	if enabled, err := protoRevDec.ProtoRevKeeper.GetProtoRevEnabled(cacheCtx); err != nil || !enabled {
+		return next(ctx, tx, simulate)
+	}
+
+	// Get the max number of pools to iterate through
+	maxPoolsToIterate, err := protoRevDec.ProtoRevKeeper.GetMaxPools(cacheCtx)
+	if err != nil {
+		next(ctx, tx, simulate)
+	}
+
+	// Find routes for every single pool that was swapped on (up to maxPoolsToIterate pools per tx)
+	swappedPools := ExtractSwappedPools(tx)
+	tradeErr := error(nil)
+	for index, swap := range swappedPools {
+		if uint64(index) >= maxPoolsToIterate {
+			break
+		}
+		// If there was is an error executing the trade, break and set tradeErr
+		if err := protoRevDec.ProtoRevKeeper.ProtoRevTrade(cacheCtx, swap); err != nil {
+			tradeErr = err
+			break
+		}
+	}
+	// If there was no error, write the cache context to the main context
+	if tradeErr == nil {
+		write()
+		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+	}
```

Comment:
As of now, this current logic will track if an error occurs when attempting to execute a trade (this should never happen, if it occurs it means something is wrong with protorev logic), and if one occurs it doesn't write anything to the main ctx for that posthandler's run (even if a previous arb executed for the first pool in a user's swap).

We could however change this logic such that if an error occurs, we refresh our cacheCtx and continue going. This would allow for the case where a user swaps against 2 pools, the first pool results in an error, but we successfully arb the second. This code is below:

`		// If there was no error executing the trade, then write the cache context to the main context,
		// emit events, and create a new cache context with an infinite gas meter from the main context
		executed, err := protoRevDec.ProtoRevKeeper.ProtoRevTrade(cacheCtx, swap)
		if err == nil && executed {
			write()
			ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
			cacheCtx, write = ctx.CacheContext()
			cacheCtx = cacheCtx.WithGasMeter(sdk.NewInfiniteGasMeter())
		} else if err != nil {
			// If there was an error executing the trade, then create a
			// new cache ctx with an infinite gas meter from the main context
			cacheCtx, write = ctx.CacheContext()
			cacheCtx = cacheCtx.WithGasMeter(sdk.NewInfiniteGasMeter())
		}`

Reply:
My preference is to keep the currently implemented logic because the error shouldn't happen. I would want to add a logger when `err != nil` though so that we could more easily identify if an error occurs while it's live.